[DOC-757] Revamp schema workaround page in voyager docs (#27137)
yugabyte/yugabyte-db: docs/content/preview/yugabyte-voyager/known-issues/postgresql.md

    parent: known-issues
    weight: 101
type: docs

> rightNav:

Reviewer: Can we have the right-side nav bar to be pinned as we scroll through the page? Right now, it does not seem to do that...

Author: There was some issue with right nav, it is fixed in production Once we merge the main changes, this pr should also reflect the sticky nav.

Review limitations and implement suggested workarounds to successfully migrate data from PostgreSQL to YugabyteDB.

> Table of Content:

Reviewer: I think the current "Contents" is fine?

Author: We're thinking of removing the table of content section itself now.

> Table of Content:

Reviewer 2: ```suggestion
## Contents
```

- [Data Definition](#data-definition)
- [Data Types](#data-types)
- [Data Manipulation](#data-manipulation)
- [Functions and Operators](#functions-and-operators)
- [Indexes](#indexes)
- [Concurrency Control](#concurrency-control)
- [Extensions](#extensions)
- [Server Programming](#server-programming)
- [Postgres 12 And later features](#postgres-12-and-later-features)
- [Migration Process and Tooling Issues](#migration-process-and-tooling-issues)
- [Performance Optimizations](#performance-optimizations)

> ## Data Definition

Reviewer 2: Convention for headings everywhere in docs is to use sentence capitalization

Author: Understood!

### Tables

> #### Table inheritance is not supported

Reviewer: This heading size is a bit too small. Can we make "Data Definition" H1 and "Tables" H2 and thereby the actual issue "Table inheritance is not supported as H3?

Reviewer: One more aspect to consider is that all issues should have similar headings. 
For example "XID functions is not supported" seems to be H3, while these under "data definition>tables>" are H4. 

I know that the nesting is a bit uneven across the issues categories. For instance "Data definition" has sub-categories,  but not all other categories do. 

How about we follow this irrespective of whether there is a subcategory or not:
category - H1
sub-category - H2
issue - H3

Reviewer 2: The docs stylesheet does not support h1. h2 is the highest level inside a page. So:
category - h2
sub-category - h3
issue - h4

Changing h4 for better contrast is something that would need to be revisited separately. For now we can  add a horizontal rule at the end each issue to provide better separation. (example in next comment)

Author: I see the problem, but right now, how the stylesheet is for the website and how we define h1, h2, and h3. Ideally, there can't be multiple H1s on a page because H1 is supposed to be the page title. For now, we can keep it as it is. Later, we can think about this problem as a whole. I mean, from a website improvement perspective.

**GitHub**: [Issue #5956](https://github.com/yugabyte/yugabyte-db/issues/5956)

**Description**: If you have table inheritance in the source database, it will error out in the target as it is not currently supported in YugabyteDB:

```output
ERROR: INHERITS not supported yet
```

**Workaround**: Currently, there is no workaround.

**Example**

An example schema on the source database is as follows:

```sql
CREATE TABLE public.cities (
    name text,
    population real,
    elevation integer
);

CREATE TABLE public.capitals (
    state character(2) NOT NULL
)
INHERITS (public.cities);
```

> 

Reviewer 2: ```suggestion
---

```

Author: Understood!

Author: I've added those breaks after each issue.

    text   jsonb,
    text1  jsonb
);

CREATE INDEX gin_multi_on_json
    ON public.test_gin_json USING gin (text, text1);
```

---

## Concurrency control

### Advisory locks is not yet implemented

**GitHub**: [Issue #3642](https://github.com/yugabyte/yugabyte-db/issues/3642)

**Description**: YugabyteDB does not support PostgreSQL advisory locks (for example, pg_advisory_lock, pg_try_advisory_lock). Any attempt to use advisory locks will result in a "function-not-implemented" error as per the following example:

```sql
yugabyte=# SELECT pg_advisory_lock(100), COUNT(*) FROM cars;
```

```output
ERROR:  advisory locks feature is currently in preview
HINT:  To enable this preview feature, set the GFlag ysql_yb_enable_advisory_locks to true and add it to the list of allowed preview flags i.e. GFlag allowed_preview_flags_csv. If the app doesn't need strict functionality, this error can be silenced by using the GFlag yb_silence_advisory_locks_not_supported_error. See https://github.com/yugabyte/yugabyte-db/issues/3642 for details
```

**Workaround**: Implement a custom locking mechanism in the application to coordinate actions without relying on database-level advisory locks.

---

> ### Events Listen / Notify

Reviewer 3: Lets move this one to `Server Programming` as we discussed in the doc. cc: @makalaaneesh

Author: Moved! Thanks for review!

```output
ERROR: invalid type name "employees.salary%TYPE" (SQLSTATE 42601)
```

**Workaround**: Fix the syntax to include the actual type name instead of referencing the type of a column.

**Example**

An example schema on the source database is as follows:

```sql
CREATE TABLE public.employees (
    employee_id integer NOT NULL,
    employee_name text,
    salary numeric
);

CREATE FUNCTION public.get_employee_salary(emp_id integer) RETURNS numeric
    LANGUAGE plpgsql
    AS $$
DECLARE
    emp_salary employees.salary%TYPE;  -- Declare a variable with the same type as employees.salary
BEGIN
    SELECT salary INTO emp_salary
    FROM employees
    WHERE employee_id = emp_id;

    RETURN emp_salary;
END;
$$;
```

Suggested change to CREATE FUNCTION is as follows:

```sql
CREATE FUNCTION public.get_employee_salary(emp_id integer) RETURNS numeric
    LANGUAGE plpgsql
    AS $$
DECLARE
    Emp_salary NUMERIC;  -- Declare a variable with the same type as employees.salary
BEGIN
    SELECT salary INTO emp_salary
    FROM employees
    WHERE employee_id = emp_id;

    RETURN emp_salary;
END;
$$;
```

---

> ## Postgres 12 And later features

Reviewer 2: ```suggestion
## PostgreSQL 12 and later features
```

### PostgreSQL 12 and later features

**GitHub**: Issue [#25575](https://github.com/yugabyte/yugabyte-db/issues/25575)

> **Description**: If any of these PostgreSQL features for version 12 and later are present in the source schema, the import schema step on the target YugabyteDB will fail as YugabyteDB is currently PG11 compatible.

Reviewer 2: ```suggestion
**Description**: If any of the following PostgreSQL 12 and later features are present in the source schema, the import schema step on the target YugabyteDB will fail.
```

- [JSON Constructor functions](https://www.postgresql.org/about/featurematrix/detail/395/) - `JSON_ARRAY_AGG`, `JSON_ARRAY`, `JSON_OBJECT`, `JSON_OBJECT_AGG`.
- [JSON query functions](https://www.postgresql.org/docs/17/functions-json.html#FUNCTIONS-SQLJSON-TABLE) - `JSON_QUERY`, `JSON_VALUE`, `JSON_EXISTS`, `JSON_TABLE`.
- [IS JSON predicate clause](https://www.postgresql.org/about/featurematrix/detail/396/).
- Any Value [Aggregate function](https://www.postgresql.org/docs/16/functions-aggregate.html#id-1.5.8.27.5.2.4.1.1.1.1) - `any_value`.
- [COPY FROM command with ON_ERROR](https://www.postgresql.org/about/featurematrix/detail/433/) option.
- [Non-decimal integer literals](https://www.postgresql.org/about/featurematrix/detail/407/).
- [Non-deterministic collations](https://www.postgresql.org/docs/12/collation.html#COLLATION-NONDETERMINISTIC).
- [COMPRESSION clause](https://www.postgresql.org/docs/current/sql-createtable.html#SQL-CREATETABLE-PARMS-COMPRESSION) in TABLE Column for TOASTing method.
- [CREATE DATABASE options](https://www.postgresql.org/docs/15/sql-createdatabase.html) (locale, collation, strategy, and oid related).

> Apart from these, the following issues are supported in YugabyteDB [v2.25](/preview/releases/ybdb-releases/v2.25), which supports PostgreSQL 15.

Reviewer 2: I'm not sure what this means. Do you mean these will also fail, unless you are importing to 2.25+?

eg:
"In addition, if any of the following PostgreSQL features are present in the source schema, the import schema step on the target YugabyteDB will fail, unless you are importing to YugabyteDB v2.25 (which supports PG15).

Author: @makalaaneesh can you please help here?

Reviewer 3: yes correct.  This makes sense.

> "In addition, if any of the following PostgreSQL features are present in the source schema, the import schema step on the target YugabyteDB will fail, unless you are importing to YugabyteDB v2.25 (which supports PG15).

Author: I've updated this sentence with what Dwight suggested.
